views-core/src/main/java/io/micronaut/views/ViewsFilter.java

```diff
@@ -113,26 +113,40 @@ public int getOrder() {
                     Optional<String> optionalView = resolveView(route, body);
 
                     if (optionalView.isPresent()) {
-
                         MediaType type = route.getValue(Produces.class, MediaType.class)
-                                .orElse((route.getValue(View.class).isPresent() || body instanceof ModelAndView) ? MediaType.TEXT_HTML_TYPE : MediaType.APPLICATION_JSON_TYPE);
-                        Optional<ViewsRenderer> optionalViewsRenderer = beanLocator.findBean(ViewsRenderer.class,
-                                new ProducesMediaTypeQualifier<>(type));
+                          .orElse((route.getValue(View.class).isPresent() ||
+                            body instanceof ModelAndView) ?
+                            MediaType.TEXT_HTML_TYPE :
+                            MediaType.APPLICATION_JSON_TYPE);
+                        Optional<BaseViewsRenderer> optionalViewsRenderer = beanLocator.findBean(
+                          BaseViewsRenderer.class,
+                          new ProducesMediaTypeQualifier<>(type));
 
                         if (optionalViewsRenderer.isPresent()) {
-                            ViewsRenderer viewsRenderer = optionalViewsRenderer.get();
+                            BaseViewsRenderer viewsRenderer = optionalViewsRenderer.get();
                             Map<String, Object> model = populateModel(request, viewsRenderer, body);
                             ModelAndView<Map<String, Object>> modelAndView = processModelAndView(request,
-                                    optionalView.get(),
-                                    model);
+                              optionalView.get(),
+                              model);
                             model = modelAndView.getModel().orElse(model);
                             String view = modelAndView.getView().orElse(optionalView.get());
-                            if (viewsRenderer.exists(view)) {
 
-                                Writable writable = viewsRenderer.render(view, model, request);
+                            if (viewsRenderer.exists(view)) {
                                 response.contentType(type);
-                                ((MutableHttpResponse<Object>) response).body(writable);
-                                return Flowable.just(response);
+
+                                if (viewsRenderer instanceof AsyncViewsRenderer) {
+                                    // it's an async renderer
+                                    //noinspection unchecked
+                                    return ((AsyncViewsRenderer) viewsRenderer).render(
+                                      view, model, request, (MutableHttpResponse<Writable>) response);
```

Comment:
if a given `viewsRenderer` is an `instanceof` `AsyncViewsRenderer`, it is now expected to return a `Flowable`, which allows the renderer to control the async flow